### DISCUSSION TITLE
Is there a wait flag?

### DISCUSSION THREAD

**Discussion by alper** [helix-editor/helix#11693] [Troubleshooting]
### Summary

I was looking to see why `chezmoi` does not work with hx as my editor but it seems there is no `--wait`/`-f` flag?

Any chance how to work around this?

### Platform

macOS

### Terminal Emulator

kitty

### Installation Method

brew

### Helix Version

helix 24.7 (079f5442)

**Comment by TornaxO7** [comment]
May I ask what you'd expect of a `--wait`/`-f` flag?

**Reply by alper** [reply]
Hi, there seems to be the thing where the editor won't return to the calling program until we're done, like described here: https://stackoverflow.com/questions/68975299/why-should-i-use-wait-when-selecting-my-default-editor-in-git

Chezmoi for instance opens a temporary file in $EDITOR and then when you're done editing does something with the result.

**Reply by alper** [reply]
It's kinda annoying having to change my editor to `vim` just to be able to use `chezmoi` which otherwise fails to pick up any changes.

**Reply by gabydd** [reply]
You shouldn't need wait with helix, chezmoi should spawn the process and then when you quit helix it will know the process exited( https://github.com/twpayne/chezmoi/blob/master/internal/cmd/config.go#L2429 ), this is what happens when you edit a git commit message for example and presumably is what is happening when using vim with chezmoi

**Reply by alper** [reply]
Yes, I agree that it does not make any sense but my experience has been that chezmoi is pretty unusable with hx at the moment. See this video where the change is just thrown away.

https://github.com/user-attachments/assets/cbc4deae-50cb-4e61-9634-bb1a5efccd7d

**Reply by twpayne** [reply]
Author of chezmoi here. Helix is fantastic and I want chezmoi and Helix to work together.

As far as I can tell https://github.com/helix-editor/helix/discussions/11693#discussioncomment-10652318 is correct, that is:

1. chezmoi runs `hx $FILENAME` and waits for `hx` to return.
2. `hx` only returns when the user quits `hx`.
 
So there is no need for a `--wait` flag. `hx` already waits by default.

I suspect that the absence of any diff after editing in the video in https://github.com/helix-editor/helix/discussions/11693#discussioncomment-10652718 is due to other reasons. For example, is it possible that @alper's chezmoi config on that machine ignores `~/.gitignore_global`?

**Reply by alper** [reply]
No, it's not ignored. If I edit the file with `code` or edit it with hx bare then I get a diff.

See here there's a clear difference in behaviour between whether I launch hx or whether chezmoi launches it:

https://github.com/user-attachments/assets/5b06a57d-1f35-4da8-a149-978ce40b84dc

**Reply by twpayne** [reply]
I'm unable to reproduce this.

I put the following in my chezmoi config (`~/.config/chezmoi/chezmoi.toml`):

```toml
[edit]
    command = "hx"
```

I then run a command like `chezmoi edit` or `chezmoi edit-config` and chezmoi opens Helix and waits for Helix to terminate.

Are you able to come up with a minimal example that enables someone else to reproduce the problem that you're seeing?

**Reply by alper** [reply]
It's ~~wild~~ odd because while I'm editing my `ps` in another tab shows this (which is exactly what I would be expecting):

```
❯ ps -A | grep chezmoi                                                                                [🕙 30 22:39] 
17826 ttys006    0:00.00 grep --color=auto chezmoi
17822 ttys008    0:00.08 chezmoi edit .gitignore_global
17823 ttys008    0:00.03 /opt/homebrew/bin/hx /var/folders/xg/3cb3b00x4gldk0j0hdy82qqr0000gn/T/chezmoi-edit3860135532/.gitignore_global
```

But when I quit, the diff is still empty.

**Reply by twpayne** [reply]
It's not clear from your response what part is "wild". If you're wondering where the `/var/folders/...` path comes from then please read up on [`chezmoi edit`](https://www.chezmoi.io/reference/commands/edit/), and particularly:

> If the operating system supports hard links, then the edit command invokes the editor with filenames which match the target filename, unless the `edit.hardlink` configuration variable is set to `false` [or] the `--hardlink=false` command line flag is set.

Are you able to reproduce what you see on any other system outside your own? If so, how?

**Reply by alper** [reply]
Not sure how to create a minimal example. I tried this and that does work:
```
chezmoi edit .gitignore_global --hardlink=0
```

I found this online:
> A lot of editors perform atomic writes by writing to a temporary file and then moving that file over the top of the file you are editing. This means you wipe out the hard link
https://superuser.com/a/1530003

**Reply by alper** [reply]
And it says: https://github.com/helix-editor/helix/issues/11279

(I think those are the exact reproduction steps that happen here because we are editing a hard link to the file.)

**Reply by twpayne** [reply]
Thank you, that makes total sense.

So, if I understand correctly, #11340 fixes the problem and should be in the next release of Helix.

**Reply by alper** [reply]
Curious why it didn't happen for you. Do you have a nightly build of helix?

**Reply by alper** [reply]
I have: `helix 24.7 (079f5442)`

**Reply by twpayne** [reply]
I also have `helix 24.7 (079f5442)` on macOS, but I was focusing on chezmoi running Helix correctly and waiting for it to finish, not on Helix not handling hardlinks correctly.
